src/projectDB/projectversioning.cpp

```diff
@@ -0,0 +1,257 @@
+#include <boost/filesystem.hpp>
+#include <regex>
+
+#include "connection.h"
+#include "cursor.h"
+#include "mzUtils.h"
+#include "projectversioning.h"
+
+namespace bfs = boost::filesystem;
+
+namespace ProjectVersioning {
+
+/**
+ * Update this for every release where DB version changes. Intermediate
+ * releases using the same DB version as the last mentioned release need not be
+ * added.
+ */
+map<Version, int> appDbVersionMap = {
+    {Version("0.6.0"), 0},
+};
+
+/**
+ * Update this for every release where DB format changes. Please only add the
+ * absolute minimum change that would be needed for moving from one database
+ * format to the next. And DO NOT assume that any of the tables that should be
+ * transformed already exist.
+ */
+map<int, string> dbVersionUpgradeScripts = {};
+
+////////////////////////////////////////////////////////////////////////////////
+
+Version::Version(string version)
+{
+    string originalString = version;
+
+    // remove spaces from the string, if any
+    auto shuffled = std::remove_if(begin(version),
+                                   end(version),
+                                   [](unsigned char x) { return isspace(x); });
+    version.erase(shuffled, end(version));
+
+    // split on '.' character
+    vector<string> versionVec;
+    mzUtils::split(version, '.', versionVec);
+
+    // get version numnbers
+    _major = _minor = _patch = 0;
+    try {
+        if (versionVec.size() > 0)
+            _major = stoi(versionVec.at(0));
+        if (versionVec.size() > 1)
+            _minor = stoi(versionVec.at(1));
+        if (versionVec.size() > 2)
+            _patch = stoi(versionVec.at(2));
+    } catch(invalid_argument) {
+        cerr << "Error: unexpected version string format - \""
+             << originalString
+             << "\""
+             << endl;
+    }
+}
+
+string Version::toString() const
+{
+    return to_string(_major)
+           + '.'
+           + to_string(_minor)
+           + '.'
+           + to_string(_patch);
+}
+
+ostream& operator << (ostream& os, const Version& v)
+{
+    os << v.toString();
+    return os;
+}
+
+void Version::upMajor()
+{
+    ++_major;
+}
+
+void Version::upMinor()
+{
+    ++_minor;
+}
+
+void Version::upPatch()
+{
+    ++_patch;
+}
+
+bool Version::downMajor()
+{
+    if (_major != 0) {
+        --_major;
+        return true;
+    }
+    return false;
+}
+
+bool Version::downMinor()
+{
+    if (_minor != 0) {
+        --_minor;
+        return true;
+    }
+    return false;
+}
+
+bool Version::downPatch()
+{
+    if (_patch != 0) {
+        --_patch;
+        return true;
+    }
+    return false;
+}
+
+int Version::_compare(const Version& other) const
+{
+    int majorDiff = _major - other.major();
+    int minorDiff = _minor - other.minor();
+    int patchDiff = _patch - other.patch();
+    return majorDiff + minorDiff + patchDiff;
```

Comment:
This method has a basic flaw. Comparing version `0.6.1` and `0.7.0` will yield 0 (indicating they are equal when they are not), because `minorDiff` will be -1 and `patchDiff` will be 1, cancelling each other out. Will make this correction